src/ExtUvLoop.php

```diff
@@ -0,0 +1,306 @@
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function run()
+    {
+        $this->running = true;
+
+        while ($this->running) {
+            $this->futureTickQueue->tick();
+
+            $hasPendingCallbacks = !$this->futureTickQueue->isEmpty();
+            $wasJustStopped = !$this->running;
+            $nothingLeftToDo = !$this->readStreams
+                && !$this->writeStreams
+                && !$this->timers->count()
+                && $this->signals->isEmpty();
+
+            // Use UV::RUN_ONCE when there are only I/O events active in the loop and block until one of those triggers,
+            // otherwise use UV::RUN_NOWAIT.
+            // @link http://docs.libuv.org/en/v1.x/loop.html#c.uv_run
+            $flags = \UV::RUN_ONCE;
+            if ($wasJustStopped || $hasPendingCallbacks) {
+                $flags = \UV::RUN_NOWAIT;
+            } elseif ($nothingLeftToDo) {
+                break;
+            }
+
+            \uv_run($this->uv, $flags);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function stop()
+    {
+        $this->running = false;
+    }
+
+    private function addStream($stream)
+    {
+        if (!isset($this->streamEvents[(int) $stream])) {
+            $this->streamEvents[(int) $stream] = \uv_poll_init_socket($this->uv, $stream);
+        }
+
+        $this->pollStream($stream);
+    }
+
+    private function removeStream($stream)
+    {
+        if (!isset($this->streamEvents[(int) $stream])) {
+            return;
+        }
+
+        if (!isset($this->readStreams[(int) $stream])
+            && !isset($this->writeStreams[(int) $stream])) {
+            \uv_poll_stop($this->streamEvents[(int) $stream]);
+            \uv_close($this->streamEvents[(int) $stream]);
+            unset($this->streamEvents[(int) $stream]);
+            unset($this->flags[(int) $stream]);
+            return;
+        }
+
+        $this->pollStream($stream);
+    }
+
+    private function pollStream($stream)
+    {
+        if (!isset($this->streamEvents[(int) $stream])) {
+            return;
+        }
+
+        $flags = 0;
+        if (isset($this->readStreams[(int) $stream])) {
+            $flags |= \UV::READABLE;
+        }
+
+        if (isset($this->writeStreams[(int) $stream])) {
+            $flags |= \UV::WRITABLE;
+        }
+
+        if (isset($this->flags[(int) $stream]) && $this->flags[(int) $stream] == $flags) {
+            return;
+        }
+
+        $this->flags[(int) $stream] = $flags;
+
+        \uv_poll_start($this->streamEvents[(int) $stream], $flags, $this->streamListener);
+    }
+
+    /**
+     * Create a stream listener
+     *
+     * @return callable Returns a callback
+     */
+    private function createStreamListener()
+    {
+        $callback = function ($event, $status, $events, $stream) {
+            if ($status !== 0) {
+                unset($this->flags[(int) $stream]);
+                $this->pollStream($stream);
+            }
```

Comment:
Do we have a test that covers this block of code? As per http://docs.libuv.org/en/v1.x/poll.html#c.uv_poll_start, it's my understanding that this would indicate an error and not a situation where the poll flags need to be updated?

Reply:
Ping @WyriHaximus, what's the status here?

Reply:
No status change

Reply:
Really unsure what to do with this so I've removed it. If we run into this we can add it again once we have a reproduceable case.

Reply:
Ok done